### PR TITLE
16 let compiler deal with different bell states

### LINKED ISSUE #16: Let compiler deal with different Bell states
The NetQASM interface provides a way for the quantum node controller to tell the host which Bell state was produced: write the type of the generated Bell state to the corresponding element in the EPR results array. 

At the moment, the Host never checks this Bell state type and always assumes that the generated Bell state is Phi+.

*Goals*
- The compiler should not assume anymore that the Bell state is always Phi+
- The compiler should make sure that the Bell state provided to the SDK layer is Phi+, or that the classical correlations are those expected from having had a Phi+ state.
- The compiler should take advantage of potential optimizations, such as the merging of Bell corrections with gates specified by the application.

### PR DESCRIPTION
**Pull request by bvdvecht**
Closes #16

### CODE REVIEW

File: netqasm/sdk/build_epr.py [hunk 1]
```diff
@@ -139,21 +150,21 @@ def deserialize_epr_keep_results(num_pairs: int, array: Array) -> List[EprKeepRe
 
 
 def deserialize_epr_measure_results(
-    num_pairs: int, array: Array
+    request: EntRequestParams, array: Array, role: EPRRole
```
Comment (QFer) [inline netqasm/sdk/build_epr.py:155]: Shouldn't deserialize_epr_keep_results get the same changes to make it consistent?

File: netqasm/sdk/builder.py [hunk 2]
```diff
@@ -1269,6 +1299,83 @@ def undef_result_element(conn: BaseNetQASMConnection, _: RegFuture):
             loop_register=index_reg,
         )
 
+    def _build_cmds_epr_keep_corrections_single_pair(
+        self, bell_state: RegFuture, qubit_reg: operand.Register
+    ) -> None:
+        with bell_state.if_eq(1):  # Phi- -> apply Z-gate
```
Comment (QFer) [inline netqasm/sdk/builder.py]: this 1, 2, 3 values, couldn't they be constants or an enum with a name?

File: netqasm/sdk/builder.py [hunk 3]
```diff
@@ -1500,7 +1622,8 @@ def _build_cmds_loop_body(
             step=step,
             loop_register=loop_register,
         )
-        self._mem_mgr.remove_active_register(loop_register)
+        if not loop_register_already_activated:
```
Comment (QFer) [inline netqasm/sdk/builder.py:1650]: The boolean check is not wrong but isn't it an idea to also include the check if reg in active_registers in remove_active_register

**Review by QFer — CHANGES_REQUESTED** [review]

File: netqasm/sdk/builder.py (hunk 3 above)
Comment (bvdvecht) [reply netqasm/sdk/builder.py:1650]: Here we want to make sure that the active state (true or false) of loop_register is the same as at the start of this method. Therefore we check `if not loop_register_already_activated` twice. I don't think it's possible with only a check inside `remove_active_register`.

File: netqasm/sdk/build_epr.py (hunk 1 above)
Comment (bvdvecht) [reply netqasm/sdk/build_epr.py:155]: Updated

File: netqasm/sdk/builder.py (hunk 2 above)
Comment (bvdvecht) [reply netqasm/sdk/builder.py]: Using the `BellState` enum now

Comment (QFer) [reply netqasm/sdk/builder.py]: You now use `BellState.PHI_MINUS.value` but `BellState.PHI_MINUS` is enough for an enum.

Comment (bvdvecht) [reply netqasm/sdk/builder.py]: `if_eq` actually requires an int (or more precisely, a `T_CValue`), so `value` is needed here

**Review by QFer — APPROVED** [review]

**Pull request merged**